src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -501,11 +501,14 @@ bool StorageReplicatedMergeTree::checkFixedGranularityInZookeeper()
 
 
 void StorageReplicatedMergeTree::waitMutationToFinishOnReplicas(
-    const Strings & replicas, const String & mutation_id) const
+    const Strings & replicas, const String & mutation_id)
 {
     if (replicas.empty())
         return;
 
+    /// We need to make sure that local queue has been synced from zookeeper and contains the entry for the
+    /// current mutation. We use local mutation status for checkig for errors.
+    queue.updateMutations(getZooKeeper());
```

Comment:
But is this part of the patch (the second definitely required) really required? There is a retry loop anyway - https://github.com/ClickHouse/ClickHouse/blob/29273d2bc1c47debb8b2098a938b7574ad7863ab/src/Storages/StorageReplicatedMergeTree.cpp#L554-L556

Reply:
Here is the scenario that we wanted to address:
1. we are on replica 2  an call waitMutationToFinishOnReplicas(replicas = {1,3})  /// local replica (2) is not on the list
2. we query ZK and see that replica 1 has already finished the mutation, so we exit the inner loop at row https://github.com/ClickHouse/ClickHouse/blob/29273d2bc1c47debb8b2098a938b7574ad7863ab/src/Storages/StorageReplicatedMergeTree.cpp#L545
3. at row https://github.com/ClickHouse/ClickHouse/blob/29273d2bc1c47debb8b2098a938b7574ad7863ab/src/Storages/StorageReplicatedMergeTree.cpp#L585 we get empty mutation_status because our current replica (2) has not synced mutations_by_znode from ZK yet. This makes checkMutationStatus() throw "Mutation {} killed (UNFINISHED)"

Reply:
> Hm, I realized, that I'm not 100% sure that it's safe to call `updateMutations` when `ReplicatedMergeTreeRestartingThread` restarts replication... (for exmaple, it's unsafe to call `pullLogsToQueue`)

Then let's just make sure that current replica is the first on the list.

Reply:
>we are on replica 2 an call waitMutationToFinishOnReplicas(replicas = {1,3}) /// local replica (2) is not on the list

But how this is possible? local replica should be always in the replicas list.

Reply:
> But how this is possible? local replica should be always in the replicas list.
Yes, currently, waitMutationToFinishOnReplicas() is called only from 1 place and local replica is always added to the list. I just want to make sure that if somebody else in the future adds a call to waitMutationToFinishOnReplicas() with a differnet list of replicas that this function will still work properly.